aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -254,6 +255,249 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
+      t);
+
+  std::vector<int64_t> out_shape;
+  std::vector<Tensor> tensors(n);
+
+  // If the first tensor is 1D of size n view it as a row vector (1, n)
+  if (_tensors[0].dim() == 1) {
+    tensors[0] = _tensors[0].view({1, -1});
+  } else if (_tensors[0].dim() == 2) {
+    tensors[0] = _tensors[0];
+    out_shape.push_back(tensors[0].size(0));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot(): the first tensor must be 1D or 2D but got ",
+        _tensors[0].dim(),
+        "D");
+  }
+
+  // If the last tensor is 1D of size n view it as a column vector (n, 1)
+  if (_tensors[n - 1].dim() == 1) {
+    tensors[n - 1] = _tensors[n - 1].view({-1, 1});
+  } else if (_tensors[n - 1].dim() == 2) {
+    tensors[n - 1] = _tensors[n - 1];
+    out_shape.push_back(tensors[n - 1].size(1));
+  } else {
+    TORCH_CHECK(
+        false,
+        "multi_dot(): the last tensor must be 1D or 2D but got ",
+        _tensors[0].dim(),
+        "D");
+  }
+
+  // Ensure middle tensors are 2D
+  for (auto i = decltype(n){1}; i < n - 1; ++i) {
+    TORCH_CHECK(
+        _tensors[i].dim() == 2,
+        "multi_dot(): tensor ",
+        i,
+        " must be 2D but got ",
+        _tensors[0].dim(),
+        "D");
+    tensors[i] = _tensors[i];
+  }
+
+  // Ensure all tensors have the same device and dtype and check
+  // that the shapes can be multiplied
+  const auto dtype = tensors[0].dtype();
+  const auto device = tensors[0].device();
+  bool has_empty_tensor = tensors[0].numel() == 0;
+  for (auto i = decltype(n){1}; i < n; ++i) {
+    TORCH_CHECK(
+        tensors[i].dtype() == dtype,
+        "multi_dot(): all tensors must have be the same dtype but tensor 0 is ",
+        dtype,
+        " and tensor ",
+        i,
+        " ",
+        tensors[i].dtype());
+    TORCH_CHECK(
+        tensors[i].device() == device,
+        "multi_dot(): all tensors must be on the same device but tensor 0 is on ",
+        device,
+        " and tensor ",
+        i,
+        " on ",
+        tensors[i].device());
+    TORCH_CHECK(
+        tensors[i - 1].size(-1) == tensors[i].size(0),
+        "multi_dot(): tensors ",
+        i - 1,
+        " and ",
+        i,
+        " with shapes ",
+        _tensors[i - 1].sizes(),
+        " and ",
+        _tensors[i].sizes(),
+        " cannot be multiplied")
+     has_empty_tensor |= tensors[i].numel() == 0;
+  }
+
+  if (out.defined()) {
+    TORCH_CHECK(
+        dtype == out.dtype(),
+        "multi_dot(): expected out tensor to have dtype ",
+        dtype,
+        " but got ",
+        out.dtype());
+    TORCH_CHECK(
+        device == out.device(),
+        "multi_dot(): expected out tensor to be on device ",
+        device,
+        " but got ",
+        out.device());
+    
+    at::native::resize_output(out, out_shape);
+    
+    // View output as 2D for computations
+    out.resize_({tensors[0].size(0), tensors.back().size(-1)});
```

Comment:
Instead of doing `resize_` here, you can do a non-inplace view `out_c = out.view(...)`, that way you don't have to `resize_` `out` again later, and `out` will automatically have the right data and the right size.

Reply:
That is a much better approach, thanks!